ext/dom/token_list.c

```diff
@@ -0,0 +1,729 @@
+		RETURN_TRUE;
+	}
+
+	/* 5. Return false. */
+	RETURN_FALSE;
+}
+
+/* https://dom.spec.whatwg.org/#dom-domtokenlist-replace */
+PHP_METHOD(Dom_TokenList, replace)
+{
+	zend_string *token, *new_token;
+	ZEND_PARSE_PARAMETERS_START(2, 2)
+		Z_PARAM_PATH_STR(token)
+		Z_PARAM_PATH_STR(new_token)
+	ZEND_PARSE_PARAMETERS_END();
+
+	/* Steps 1 - 2 */
+	if (!dom_validate_token(token) || !dom_validate_token(new_token)) {
+		RETURN_THROWS();
+	}
+
+	/* 3. If this’s token set does not contain token, then return false. */
+	dom_token_list_object *intern = TOKEN_LIST_GET_INTERNAL();
+	dom_token_list_ensure_set_up_to_date(intern);
+	HashTable *token_set = TOKEN_LIST_GET_SET(intern);
+	zval *found_token = zend_hash_find(token_set, token);
+	if (found_token == NULL) {
+		RETURN_FALSE;
+	}
+
+	/* 4. Replace token in this’s token set with newToken. */
+	ZEND_ASSERT(XtOffsetOf(Bucket, val) == 0 && "the cast only works if this is true");
+	Bucket *bucket = (Bucket *) found_token;
+	if (zend_hash_set_bucket_key(token_set, bucket, new_token) == NULL) {
+		/* It already exists, remove token instead. */
+		zend_hash_del_bucket(token_set, bucket);
+	} else {
+		Z_STR(bucket->val) = new_token;
+	}
+
+	/* 5. Run the update steps. */
+	dom_token_list_update(intern);
+
+	/* 6. Return true. */
+	RETURN_TRUE;
+}
+
+/* https://dom.spec.whatwg.org/#concept-domtokenlist-validation */
+PHP_METHOD(Dom_TokenList, supports)
+{
+	zend_string *token;
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_PATH_STR(token)
+	ZEND_PARSE_PARAMETERS_END();
+
+	/* The spec designers have designed the TokenList API with future usages in mind.
+	 * But right now, this should just always throw a TypeError because the only user is classList, which
+	 * does not define a supported token set. */
+	zend_throw_error(zend_ce_type_error, "Attribute \"class\" does not define any supported tokens");
+}
+
+PHP_METHOD(Dom_TokenList, count)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+	dom_token_list_object *intern = TOKEN_LIST_GET_INTERNAL();
+	dom_token_list_ensure_set_up_to_date(intern);
+	RETURN_LONG(zend_hash_num_elements(TOKEN_LIST_GET_SET(intern)));
+}
+
+PHP_METHOD(Dom_TokenList, getIterator)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+	zend_create_internal_iterator_zval(return_value, ZEND_THIS);
+}
+
+static void dom_token_list_it_dtor(zend_object_iterator *iter)
+{
+	zval_ptr_dtor(&iter->data);
+}
+
+static void dom_token_list_it_rewind(zend_object_iterator *iter)
+{
+	dom_token_list_it     *iterator = (dom_token_list_it *) iter;
+	dom_token_list_object *object   = Z_TOKEN_LIST_P(&iter->data);
+	zend_hash_internal_pointer_reset_ex(TOKEN_LIST_GET_SET(object), &iterator->pos);
+}
+
+/* TODO: this function exists because zend_hash_get_current_pos_ex doesn't exist. */
+static zend_always_inline HashPosition get_hash_current_pos_ex(HashTable *ht, HashPosition pos)
+{
+	ZEND_ASSERT(!HT_IS_PACKED(ht));
+	while (pos < ht->nNumUsed && Z_ISUNDEF(ht->arData[pos].val)) {
+		pos++;
+	}
+	return pos;
+}
```

Comment:
Might be useful to "upstream" into Zend as a follow-up

Reply:
Filed https://github.com/php/php-src/pull/14770